pkg/aws/services/acm/acm.go

```diff
@@ -0,0 +1,179 @@
+package acm
+
+import (
+	"fmt"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/acm"
+	"github.com/giantswarm/microerror"
+
+	"github.com/giantswarm/irsa-operator/pkg/aws/services/route53"
+	"github.com/giantswarm/irsa-operator/pkg/key"
+	"github.com/giantswarm/irsa-operator/pkg/util"
+)
+
+func (s *Service) EnsureCertificate(domain string, customerTags map[string]string) (*string, error) {
+	s.scope.Info(fmt.Sprintf("Ensuring ACM certificate for domain %q", domain))
+
+	// Check if certificate exists
+	existing, err := s.findCertificateForDomain(domain)
+	if err != nil {
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+	}
+
+	if existing != nil {
+		s.scope.Info("ACM certificate already exists")
+
+		return existing, nil
+	}
+
+	input := &acm.RequestCertificateInput{
+		DomainName: aws.String(domain),
+		Options:    &acm.CertificateOptions{},
+		Tags: []*acm.Tag{
+			{
+				Key:   aws.String(key.S3TagOrganization),
+				Value: aws.String(util.RemoveOrg(s.scope.ClusterNamespace())),
+			},
+			{
+				Key:   aws.String(key.S3TagCluster),
+				Value: aws.String(s.scope.ClusterName()),
+			},
+			{
+				Key:   aws.String(fmt.Sprintf(key.S3TagCloudProvider, s.scope.ClusterName())),
+				Value: aws.String("owned"),
+			},
+			{
+				Key:   aws.String(key.S3TagInstallation),
+				Value: aws.String(s.scope.Installation()),
+			},
+		},
+		ValidationMethod: aws.String(acm.ValidationMethodDns),
+	}
+
+	for k, v := range customerTags {
+		tag := &acm.Tag{
+			Key:   aws.String(k),
+			Value: aws.String(v),
+		}
+		input.Tags = append(input.Tags, tag)
+	}
+
+	s.scope.Info("Creating ACM certificate")
+
+	output, err := s.Client.RequestCertificate(input)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	s.scope.Info("ACM certificate created successfully")
+	return output.CertificateArn, nil
+}
+
+// IsCertificateIssued checks if an ACM certificate is issued.
+func (s *Service) IsCertificateIssued(arn string) (bool, error) {
+	s.scope.Info("Checking status of ACM certificate")
+
+	output, err := s.Client.DescribeCertificate(&acm.DescribeCertificateInput{
+		CertificateArn: aws.String(arn),
+	})
+	if err != nil {
+		return false, err
+	}
+
+	return *output.Certificate.Status == acm.CertificateStatusIssued, nil
+}
+
+// IsValidated checks wheter an ACM certificate's ownership is already validated or not.
+func (s *Service) IsValidated(arn string) (bool, error) {
+	output, err := s.Client.DescribeCertificate(&acm.DescribeCertificateInput{
+		CertificateArn: aws.String(arn),
+	})
+	if err != nil {
+		return false, err
+	}
+
+	return *output.Certificate.DomainValidationOptions[0].ValidationStatus == acm.DomainStatusSuccess, nil
```

Comment:
do we need to check if DomainValidationOptions array is > 0 otherwise this will panic?

Reply:
good catch thanks